src/Umbraco.Web/Routing/ContentFinderByNiceUrlAndTemplate.cs

```diff
@@ -1,57 +1,58 @@
 using Umbraco.Core.Logging;
 using Umbraco.Core.Models;
 using Umbraco.Core;
+using Umbraco.Core.Configuration;
 
 namespace Umbraco.Web.Routing
 {
-	/// <summary>
-	/// Provides an implementation of <see cref="IContentFinder"/> that handles page nice urls and a template.
-	/// </summary>
-	/// <remarks>
-	/// <para>Handles <c>/foo/bar/template</c> where <c>/foo/bar</c> is the nice url of a document, and <c>template</c> a template alias.</para>
-	/// <para>If successful, then the template of the document request is also assigned.</para>
-	/// </remarks>
+    /// <summary>
+    /// Provides an implementation of <see cref="IContentFinder"/> that handles page nice urls and a template.
+    /// </summary>
+    /// <remarks>
+    /// <para>Handles <c>/foo/bar/template</c> where <c>/foo/bar</c> is the nice url of a document, and <c>template</c> a template alias.</para>
+    /// <para>If successful, then the template of the document request is also assigned.</para>
+    /// </remarks>
     public class ContentFinderByNiceUrlAndTemplate : ContentFinderByNiceUrl

```

Comment:
I would say fall back to the original template, it's also what we do when you do ?altTemplate=Bla when there is no template "Bla". 
When you remove `<notFound assembly="umbraco" type="SearchForTemplate"/>` from the 4040handlers.config you get a 404, so if people prefer that then they can make it 404 if necessary.